### ISSUE TITLE
`ESLintUtils.RuleCreator.withoutDocs` typing isn't compatible with `defineConfig`

### ISSUE THREAD

**Issue by juhort** [typescript-eslint/typescript-eslint#12148]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

utils

### Repro Code

Here's a super simple custom rule created using `@typescript-eslint/utils`:

```TypeScript
import { ESLintUtils } from "@typescript-eslint/utils";

const noBarRule = ESLintUtils.RuleCreator.withoutDocs({
  name: "no-bar",
  meta: {
    type: "problem",
    schema: [],
    messages: {
      noBar: 'Identifier name "bar" is not allowed.',
    },
  },
  defaultOptions: [],
  create(context) {
    return {
      Identifier(node) {
        if (node.name === "bar") {
          context.report({ node, messageId: "noBar" });
        }
      },
    };
  },
});

export default noBarRule;
```

### ESLint Config

```javascript
// eslint.config.ts

import { defineConfig } from "eslint/config";
import tseslint from "typescript-eslint";
import noBarRule from "./rules/no-bar.ts";

export default defineConfig(tseslint.configs.recommended, {
  plugins: {
    example: {
      rules: {
        "no-bar": noBarRule,
      },
    },
  },
  rules: {
    "example/no-bar": "error",
  },
});
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "module": "nodenext",
    "target": "esnext",
    "types": [],

    // Other Outputs
    "sourceMap": true,
    "declaration": true,
    "declarationMap": true,

    // Stricter Typechecking Options
    "noUncheckedIndexedAccess": true,
    "exactOptionalPropertyTypes": true,

    // Recommended Options
    "strict": true,
    "jsx": "react-jsx",
    "verbatimModuleSyntax": true,
    "isolatedModules": true,
    "noUncheckedSideEffectImports": true,
    "moduleDetection": "force",
    "skipLibCheck": true,

    "noEmit": true,
    "allowImportingTsExtensions": true
  }
}
```

### Expected Result

There should _not_ be any errors in the config.

In fact there's no error when using `tseslint.config` instead of `defineConfig`, but `tseslint.config` is deprecated.

### Actual Result

This config produces the following error at `"no-bar"`:

```
Type 'RuleModule<"noBar", [], unknown, RuleListener>' is not assignable to type 'RuleDefinition<RuleDefinitionTypeOptions>'.
  Types of property 'create' are incompatible.
    Type '(context: Readonly<RuleContext<"noBar", []>>) => RuleListener' is not assignable to type '(context: RuleContext<{ LangOptions: LanguageOptions; Code: SourceCode<{ LangOptions: LanguageOptions; RootNode: unknown; SyntaxElementWithLoc: unknown; ConfigNode: unknown; }>; RuleOptions: unknown[]; Node: unknown; MessageIds: string; }>) => RuleVisitor'.
      Types of parameters 'context' and 'context' are incompatible.
        Type 'RuleContext<{ LangOptions: LanguageOptions; Code: SourceCode<{ LangOptions: LanguageOptions; RootNode: unknown; SyntaxElementWithLoc: unknown; ConfigNode: unknown; }>; RuleOptions: unknown[]; Node: unknown; MessageIds: string; }>' is missing the following properties from type 'Readonly<RuleContext<"noBar", []>>': parserOptions, parserPath, getAncestors, getDeclaredVariables, and 6 more.
```

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `typescript-eslint`             | `8.57.1` |
| `TypeScript`                           | `5.9.3` |
| `ESLint`                               | `10.0.3` |
| `node`                                 | `25.7.0` |
